src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/EETypeNode.GCDesc.cs

```diff
@@ -0,0 +1,200 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+using Internal.TypeSystem;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    partial class EETypeNode
+    {
+        private int GCDescSize
+        {
+            get
+            {
+                if (!_constructed || _type.IsGenericDefinition)
+                    return 0;
+
+                if (_type.IsArray)
+                {
+                    TypeDesc elementType = ((ArrayType)_type).ElementType;
+                    if (elementType.IsObjRef)
+                    {
+                        // For efficiency this is special cased and encoded as one serie.
+                        return 3 * _type.Context.Target.PointerSize;
+                    }
+                    else
+                    {
+                        if (elementType.IsPointer)
```

Comment:
This should just check for `elementType.IsValueType && ((DefType)elementType).ContainsPointers`